### ISSUE TITLE
[New build - EXPERIMENTAL] 20w41a: Workshop Improvements

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#3038]
Some various hotfixes, mainly logic related.

Breaking Change:
- Grating paint is now per brick, not per voxel. It means you will have to repaint all your grating!

Improvements:
- Linking gates will propagate their starting state (#2902)
- Sequence will push value on edit (#2902)
- Blueprint will now be moved to a dedicated folder (#3036)

Bugfixes:
#3030 Cannot build blocks on large pistons.
#3029 Workshop upload unsuccessful when name end with spaces.
#3023 Half rounded brick wrong scale.
#3020 #2660 #2998 Collisions issue with long Lattice Girder Large block.
#2983 #2980 Logical gate dont update correctly if several inputs are updated in the same frame.
#2960 Pressing ESC to quit also open the main menu.
#2902 NAND gate incorrect state with no input.
#2952 Copy-paste: Grating not copying paint

**Comment by tsunamayo** [comment]
Build is up! Thanks

**Comment by Kaiser-Indrasil** [comment]
Rotators and hinges work as intended now, good job!
I still have some issues pertaining to somewhat different collision detection in my turrets but I expect that they simply need to have a bit of cluttering geometry removed.

I noticed that we can't use manned turrets because there's no Axis Setting in the rotator's and hinge's menu anymore.

By the way, what is this?
![image](https://user-images.githubusercontent.com/63756057/95183530-042a6500-07be-11eb-964f-161286f49209.png)

**Comment by Kaiser-Indrasil** [comment]
I'm afraid I was overly optimistic in my estimations...
Please take a look at the video below.
If you look at the 2 turrets at the top front of the ship, you'll see that they track the target very sparsely, even if it's in their field of view.
Not to mention, these are the only 2 turrets that work out of total 7...
https://youtu.be/CFDAVjKOuDw

Also, looks like the turrets are even more inaccurate now, as they display a very sluggish target tracking:
[uJ62TOpNF5.zip](https://github.com/tsunamayo/Starship-EVO/files/5333146/uJ62TOpNF5.zip)

**Comment by Kaiser-Indrasil** [comment]
Okay, I managed to pinpoint what's exactly wrong with the Hinges right now.
When you set their current angle, it's between -90 and 90 degrees.
Meanwhile, when you set the angle with a Sequencer gate, it's actually between 0 and 180 degrees.

This results in a weird behaviour where, no matter what angle you set it to, and whether or not you use the resting position option, once you get to the yoke, the elevation hinges on the turrets just "teleport" themselves to the -90 / 0 degrees angle, facing straight down, in my case - right through the hardpoint they rest upon.

Here's the bug in action on a freshly made turret:
https://youtu.be/CyOoBrmd_tA

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil do you use sequence in combination of the turret?
Also do you reproduce straight away? 
ie you load the game, jump into your ship control, spawn an AI with F8 and see the issue.
This is what I am doing, and your turrets work just fine.
Maybe you are activating something beforehand?
Thanks

**Comment by tsunamayo** [comment]
Also, could you try the following:
Make a new ship with just a single turret, the way you did in your video, and tell me if it failed. If it failed send me the bp!
For now I just need to reproduce the issue. It is very frustrating as it is working fine over here (tried on two computer), just like for the blueprint issue...
I cant fix anything if I dont see the bug... cheers

**Comment by tsunamayo** [comment]
hi mate, so could please test this one and tell me how it goes:
[Mini Turret.zip](https://github.com/tsunamayo/Starship-EVO/files/5337608/Mini.Turret.zip)
I can indeed see that it reset to -90 the first time you activate it, but it still can track fine...
Thanks

**Comment by tsunamayo** [comment]
okay in 20w41b I fixed the issue that was pushing the turret to the wrong state on activate. Hopefully that was the root of the issue! Tell me how it is! thx

**Comment by Kaiser-Indrasil** [comment]
Okay, the barebones turret works as intended now, thanks!
But, all my normal turrets are not moving at all, or are moving, but very inconsistently, freezing for a good moment very often and clipping into themselves.
Something must be wrong with the collision detection because they used to work fine:
https://youtu.be/HUgCt8_Q0wM
Now they work like this (with no extra steps - just get into the yoke and spawn an enemy):
[P8qXc4q3b6.zip](https://github.com/tsunamayo/Starship-EVO/files/5340274/P8qXc4q3b6.zip)

Here's the blueprint I've been using in the tests:
[SoE Astero - turret test 41C.zip](https://github.com/tsunamayo/Starship-EVO/files/5340277/SoE.Astero.-.turret.test.41C.zip)

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil still works fine for me...
https://youtu.be/8zhYVdo_LB0
I just dont get it... Could someone test that ship and tell me how it goes?

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil I made a post #3051 .
By the way your camera work is so smoooth and angle are neat, I should ask you some help to shoot trailers...
thx

**Comment by tsunamayo** [comment]
btw after having a deeper look at your turret it seems they are quite complex, they have additional children entity in between, this is not supported... Are the additional children just for entity or are they actually retracting or else? Thanks

**Comment by tsunamayo** [comment]
Also regarding the test: I need you to test straight up after having loaded the game:
1) load the game
2) enter ship
3) spawn enemy with F8
4) shot at him
bug or not?

**Comment by Kaiser-Indrasil** [comment]
Yeah, these are the exact steps I took on a few tries, always with the same, buggy result.
I even used the blueprint you've posted in #3051 thinking that it might somehow differ from my local one.

The turrets are indeed complex but all subgrids they use are purely for decoration - they don't contribute to the workflow of the turret. Barebones turret looks like this:
![Starship_EVO_21qrlXVNUo](https://user-images.githubusercontent.com/63756057/95402071-b79e7100-0906-11eb-8b78-1f4da346347d.png)
and here's the same turret, just without extra sub-grids:
![Starship_EVO_im9yvGfStg](https://user-images.githubusercontent.com/63756057/95402097-c71dba00-0906-11eb-9a34-0d800a88d58e.png)
Also, thanks for the recognition for the video!
I can be of help anytime you need to shoot a trailer.

I'm quite invested in the subject of turrets ever since I've made a comprehensive tutorial about them 😄 
https://www.youtube.com/watch?v=WjOKmTmrPxk

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil do they work better without the bricks like that?

**Comment by Kaiser-Indrasil** [comment]
I've managed to get it to work like that. Still fiddling with it to pinpoint the exact bricks I need to adjust. It looks like the azimuth rotator cannot move if the elevation hinge cannot move, even if the azimuth is not obstructed itself.
![image](https://user-images.githubusercontent.com/63756057/95403146-5330e100-0909-11eb-8a2a-0b12ed1420c2.png)

**Comment by Kaiser-Indrasil** [comment]
I've managed to get them to move around using the new Current Angle mechanic but they are still completely static when I'm targeting an enemy.

Maybe it's related to the new turret collision detection system?

**Comment by tsunamayo** [comment]
No the turret system is just not made to work with a children entity in the middle between the two axis.
I will try to prevent the player from doing that for the time being, while waiting to support that (if it is possible).
Also I fixed a bug - but your turret are still stuck. I guess it was working for me because of the bug and a few other things.

**Comment by Kaiser-Indrasil** [comment]
No no, I didn't place any child entities between turret axes, just on top of them.

I learned the importance of that through 2 whole days spent on rebuilding the turrets from the ground up because before there were child entities between the axes X'D

**Comment by tsunamayo** [comment]
Mate your turrets are so confusing lol, so it just there for show? No I will collide against those entities, which I was not doing before. So if you used entity vs entity clipping (bug still not fixed), they wont work. I will try to get rid of them to see if it improves.

**Comment by Kaiser-Indrasil** [comment]
Yeah, I might have to axe those entire turret designs and settle for something else, since this is a fringe case and the vast majority of the community seems to benefit from the latest changes to turrets and mechanisms.

**Comment by tsunamayo** [comment]
ho yeah I know what is going wrong, yes having those entity will trigger a lot of collisions. It would require a dev to deal with them - I need to think a bit more to see how big it is...
